### PR TITLE
feat: enhance stream density function

### PR DESCRIPTION
**Pull request by jim-wang-intel**
Now the increment of pipeline numbers can be larger if the system is high power and performed well:
   The increment number is **dynamically adjusted** based the actual average FPS per stream and the target FPS.

   The searching policy is greedy to attempt to get to the maximum pipelines first and then decrement by 1 at a time
   to find the maximum number of pipelines reaching the target FPS.

  Closes: #51

_The PR review is to check for sustainability and correctness.  Sustainability is actually more business critical as correctness is largely tested into the code over time.   Its useful to keep in mind that SW often outlives the HW it was written for and engineers move from job to job so it is critical that code developed for Intel be supportable across many years.  It is up to the submitter and reviewer to look at the code from a perspective of what if we have to debug this 3 years from now after the author is no longer available and defect databases have been lost.  Yes, that happens all the time when we are working with time scales of more than 2 years.  When reviewing your code it is important to look at it from this perspective._

Author Mandatory (to be filled by PR Author/Submitter)
------------------------------------------------------
- Developer who submits the Pull Request for merge is required to mark the checklist below as applicable for the PR changes submitted.  
- Those checklist items which are not marked are considered as not applicable for the PR change.  
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the developers for Inner Source Development Model (ISDM) compliance

### PULL DESCRIPTION
  Now the increment of pipeline numbers can be larger if the system is high power and performed well:
   The increment number is **dynamically adjusted** based the actual average FPS per stream and the target FPS.

   The searching policy is greedy to attempt to get to the maximum pipelines first and then decrement by 1 at a time
   to find the maximum number of pipelines reaching the target FPS.

### REFERENCES
Reference URL for issue tracking (Github): **\<URL to be filled>**

- [x] Added label to the Pull Request for easier discoverability and search

### CODE MAINTAINABILITY
- [x] **_Commit Message meets guidelines as indicated in the URL\*_**
	- https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md
- [x] **_Every commit is a single defect fix and does not mix feature addition or changes\*_**
- [ ] Added required new tests relevant to the changes
	- [ ] PR contains URL links to functional tests executed with the new tests 
- [ ] Updated Documentation as relevant to the changes
- [ ] Updated Build steps/commands changes as relevant
- [ ] PR change contains code related to security
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide description)
- [ ] Specific instructions or information for code reviewers (If any):

Maintainer Mandatory (to be filled by PR Reviewer/Approving Maintainer)
-----------------------------------------------------------------------
- Maintainer who approve the Pull Request for merge are required to mark the checklist below as appropriate for the PR change reviewed as key proof of attestation indicating reasons for merge. 
- Those checklist items which are not marked are considered as not applicable for the PR change. 
- Items marked with an asterisk suffix are mandatory items to check.

### QUALITY CHECKS
- [ ] Architectural and Design Fit
- [ ] **_Quality of code (At least one should be checked as applicable)\*_**
	- [ ] Commit Message meets guidelines
	- [ ] PR changes adhere to industry practices and standards
	- [ ] Adopted domain specific coding standards ([SBL Coding Standards](https://github.com/tianocore-docs/Docs/raw/master/Specifications/CCS_2_1_Draft.pdf))
	- [ ] Error and exception code paths implemented correctly
	- [ ] Code reviewed for domain or language specific anti-patterns
	- [ ] Code is adequately commented
	- [ ] Code copyright is correct
	- [ ] Tracing output are minimized and logic
	- [ ] Confusing logic is explained in comments
	- [ ] Commit comment can be used to design a new test case for the changes
- [ ] **_Test coverage shows adequate coverage with required CI functional tests pass on all supported platforms\*_**
- [ ] **_Static code scan report shows zero critical issues\*_**

### CODE REVIEW IMPACT
- Summary of Defects Detected in Code Review: **\<%P1*xx,P2*xx,P3*xx,P4*xx%>** \
Note P1/P2/P3/P4 denotes severity of defects found (Showstopper/High/Medium/Low) and xx denotes number of defects found

# _Code must act as a teacher for future developers_

### CODE REVIEW

**Comment by jim-wang-intel** [comment]
will change the way on handling the pipelines when we are doing decrementing

File: benchmark-scripts/test_stream_density.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply benchmark-scripts/test_stream_density.sh:1]: this is the one i'm using to test

**Review by vli11 — APPROVED** [review]
LGTM

**Review by marcpfuller — APPROVED** [review]

**Review by marcpfuller — APPROVED** [review]

File: benchmark-scripts/benchmark.sh [hunk 2]
```diff
@@ -22,13 +22,15 @@ show_help() {
...
            [ --barcode_disabled | --barcode [BARCODE_INTERVAL] ]
-           [realsense_enabled]
+           [--realsense_enabled]
 
          Note: 
           1. dgpu.x should be replaced with targetted GPUs such as dgpu (for all GPUs), dgpu.0, dgpu.1, etc
           2. filesrc will utilize videos stored in the sample-media folder
           3. Set environment variable STREAM_DENSITY_MODE=1 for starting single container stream density testing
           4. Set environment variable RENDER_MODE=1 for displaying pipeline and overlay CV metadata
+          5. Stream density can take two numbers: first one is for target fps and
```
Comment (brian-intel) [inline benchmark-scripts/benchmark.sh:32]: suggest changing "two numbers" to "two parameters" and add that fps is a float value

File: benchmark-scripts/test_stream_density.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (brian-intel) [reply benchmark-scripts/test_stream_density.sh:1]: Ok I read it wrong and thought it was a copy of the stream_density.sh. having this makes more sense now

File: benchmark-scripts/test_stream_density.sh [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+statusCode=$?
+if [ $statusCode -ne 0 ]
+then
+  echo "test failed: expecting get status code 0 but found $statusCode"
+else
+  # Max stream density achieved for target FPS 12 is at least 1
+  res=$(grep -i -Eo "Max stream density achieved for target FPS ([0-9]+(.[0-9]+)*) is ([0-9])+" ../results/stream_density.log  | awk -F ' ' '{print $10}')
+
+  if [ -z "${res}" ]; then
+    echo "test fialed: maximum pipeline numbers not found"
```
Comment (brian-intel) [inline benchmark-scripts/test_stream_density.sh:27]:
```suggestion
    echo "test failed: maximum pipeline numbers not found"
```

File: benchmark-scripts/test_stream_density.sh [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
```
Comment (brian-intel) [inline benchmark-scripts/test_stream_density.sh:52]:
```suggestion
    echo "test failed: maximum pipeline numbers not found"
```

File: configs/dlstreamer/framework-pipelines/stream_density.sh [hunk 5]
```diff
@@ -5,17 +5,80 @@
...
+	waitForChildPidKilled=0
+	if [ -z "$childPids" ]
+	then
+		echo "for parent pid $pidToKill, there is no child pids to kill"
+	else
+		echo "parent pid $pidToKill with childPids $childPids to be killed"
+		waitForChildPidKilled=1
+	fi
+
+	pkill -P $pidToKill
```
Comment (brian-intel) [inline configs/dlstreamer/framework-pipelines/stream_density.sh:24]: Recommend adding a short comment like "# Kill parent process" and another at line 48 "# make sure parent pid is gone"

**Review by brian-intel — APPROVED** [review]
LGTM

**Review by jstein13 — APPROVED** [review]

**Pull request merged**